node/pkg/watchers/evm/watcher.go

```diff
@@ -574,12 +617,24 @@ func (w *Watcher) Run(ctx context.Context) error {
 				w.pendingMu.Lock()
 
 				blockNumberU := ev.Number.Uint64()
-				atomic.StoreUint64(&currentBlockNumber, blockNumberU)
-				atomic.StoreUint64(&w.latestFinalizedBlockNumber, blockNumberU)
+				if ev.Safe {
+					atomic.StoreUint64(&currentSafeBlockNumber, blockNumberU)
+				} else {
+					atomic.StoreUint64(&currentBlockNumber, blockNumberU)
+					atomic.StoreUint64(&w.latestFinalizedBlockNumber, blockNumberU)
+				}
 
 				for key, pLock := range w.pending {
+					// If this block is safe, only process messages wanting safe.
+					// If it's not safe, only process messages wanting finalized.
+					if safeBlocksSupported {
+						if ev.Safe != (pLock.message.ConsistencyLevel == vaa.ConsistencyLevelSafe) {
+							continue
+						}
+					}
+
 					var expectedConfirmations uint64
-					if w.waitForConfirmations {
+					if w.waitForConfirmations && !ev.Safe {
```

Comment:
Not sure I understand why the change to this check is needed. In fact, It may not make sense to allow `waitForConfirmations` to be set when `useFinalized` is true (and hence the only case where `Safe` could be true). Had to do some heavy thonking on this one though and I haven't finished my coffee, so I'm open to thoughts.

Reply:
We definitely don't want to add the consistency level if we are processing a message requesting safe (since adding 201 would make no sense). Technically the check may not be necessary due to the implied relationship you mentioned, but I think I would like to leave the check just to be safe / clear.